src/Core/Services/Implementations/OrganizationService.cs

```diff
@@ -184,187 +184,248 @@ public async Task ReinstateSubscriptionAsync(Guid organizationId)
         {
             throw new BadRequestException("Existing plan not found.");
         }
+        var useSecretsManager = organization.UseSecretsManager
+            ? organization.UseSecretsManager
+            : upgrade.UseSecretsManager;
 
-        var newPlan = StaticStore.PasswordManagerPlans.FirstOrDefault(p => p.Type == upgrade.Plan && !p.Disabled);
-        if (newPlan == null)
+        string paymentIntentClientSecret = null;
+        var success = true;
+        var newPlans = StaticStore.Plans.Where(p => p.Type == upgrade.Plan && !p.Disabled).ToList();
```

Comment:
I think you should include or exclude the secrets manager plan in this step. Otherwise it's included in the loop below, even when it's not required (even though you do skip it in `ValidateOrganizationUpgradeParameters`).

Also, many of the checks in the loop below only need to be done once. e.g. `UseSso` is going to be the same between the `Plan` objects. And other checks are PM-specific, like `MaxCollections` (SM doesn't have collections).

I suggest reverting the loop and leaving all the old logic in place - this covers general checks (like `UseSso`) and PM-specific checks (like `MaxCollections`). Then just add a new section for your SM-specific checks if `UseSecretsManager` is `true`, preferably in a private method.

Reply:
worked on it